packages/frontend-api/src/Model/Category/CategoryRepository.php

```diff
@@ -0,0 +1,117 @@
+
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\Query\Expr\Join;
+use Doctrine\ORM\QueryBuilder;
+use Shopsys\FrameworkBundle\Component\String\DatabaseSearching;
+use Shopsys\FrameworkBundle\Model\Category\Category;
+use Shopsys\FrameworkBundle\Model\Category\CategoryDomain;
+
+class CategoryRepository
+{
+    /**
+     * @var \Doctrine\ORM\EntityManagerInterface
+     */
+    protected EntityManagerInterface $em;
+
+    /**
+     * @param \Doctrine\ORM\EntityManagerInterface $em
+     */
+    public function __construct(EntityManagerInterface $em)
+    {
+        $this->em = $em;
+    }
+
+    /**
+     * @param string $searchText
+     * @param string $locale
+     * @param int $domainId
+     * @param int $offset
+     * @param int $limit
+     * @return \Shopsys\FrameworkBundle\Model\Category\Category[]
+     */
+    public function getVisibleCategoriesBySearchText(string $searchText, string $locale, int $domainId, int $offset, int $limit): array
+    {
+        $queryBuilder = $this->getVisibleCategoriesBySearchTextQueryBuilder($searchText, $locale, $domainId);
+
+        $queryBuilder
+            ->orderBy('ct.name')
+            ->setFirstResult($offset)
+            ->setMaxResults($limit);
+
+        return $queryBuilder->getQuery()->execute();
+    }
+
+    /**
+     * @param string $searchText
+     * @param string $locale
+     * @param int $domainId
+     * @return \Doctrine\ORM\QueryBuilder
+     */
+    protected function getVisibleCategoriesBySearchTextQueryBuilder(string $searchText, string $locale, int $domainId): QueryBuilder
+    {
+        $queryBuilder = $this->getAllVisibleByDomainIdQueryBuilder($domainId);
+
+        $this->addTranslation($queryBuilder, $locale);
+        $this->filterBySearchText($queryBuilder, $searchText);
+
+        return $queryBuilder;
+    }
+
+    /**
+     * @param string $searchText
+     * @param string $locale
+     * @param int $domainId
+     * @return int
+     */
+    public function getVisibleCategoriesBySearchTextCount(string $searchText, string $locale, int $domainId): int
+    {
+        $queryBuilder =
+            $this->getVisibleCategoriesBySearchTextQueryBuilder($searchText, $locale, $domainId)
+                ->select('COUNT(c)');
+
+        return (int)$queryBuilder->getQuery()->getSingleScalarResult();
+    }
+
+    /**
+     * @param \Doctrine\ORM\QueryBuilder $categoriesQueryBuilder
+     * @param string $locale
+     */
+    protected function addTranslation(QueryBuilder $categoriesQueryBuilder, string $locale): void
+    {
+        $categoriesQueryBuilder
+            ->addSelect('ct')
+            ->join('c.translations', 'ct', Join::WITH, 'ct.locale = :locale')
+            ->setParameter('locale', $locale);
+    }
+
+    /**
+     * @param \Doctrine\ORM\QueryBuilder $queryBuilder
+     * @param string|null $searchText
+     */
+    protected function filterBySearchText(QueryBuilder $queryBuilder, string $searchText): void
```

Comment:
I'm not happy about DRY violations... 
I see several methods copy-pasted from framework's CategoryRepository...
I'm thinking about refactoring methods in framework so they can be reused in frontend-api facade

Reply:
In repository are two problems: 
- getAllQueryBuilder adds ordering that caused me problems in counting method
- most of the methods are protected so I cannot use them

Reply:
ef3a79dbfc78d9174688ee0b361c0fec530639ce